src/Lean/Elab/Tactic/LibrarySearch.lean

```diff
@@ -0,0 +1,86 @@
+/-
+Copyright (c) 2021-2024 Gabriel Ebner and Lean FRO. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gabriel Ebner, Joe Hendrix, Scott Morrison
+-/
+import Lean.Meta.Tactic.LibrarySearch
+import Lean.Meta.Tactic.TryThis
+import Lean.Elab.Tactic.BuiltinTactic
+
+namespace Lean.Elab.LibrarySearch
+
+open Lean Meta LibrarySearch
+open Elab Tactic TryThis
+
+
+-- TODO: implement the additional options for `library_search` from Lean 3,
+-- in particular including additional lemmas
+-- with `std_exact? [X, Y, Z]` or `std_exact? with attr`.
+
+-- For now we only implement the basic functionality.
+-- The full syntax is recognized, but will produce a "Tactic has not been implemented" error.
+
+/-- Implementation of the `exact?` tactic. -/
+def exact? (tk : Syntax) (required : Option (Array (TSyntax `term)))
+   (solver : Option (TSyntax `Lean.Parser.Tactic.tacticSeq)) (requireClose : Bool) :
+    TacticM Unit := do
+  let mvar ← getMainGoal
+  let (_, goal) ← (← getMainGoal).intros
+  goal.withContext do
+    let required := (← (required.getD #[]).mapM getFVarId).toList.map .fvar
+    let tactic ←
+      match solver with
+      | none =>
+        pure (fun exfalso => solveByElim required (exfalso := exfalso) (maxDepth := 6))
+      | some t =>
+        let _ <- mkInitialTacticInfo t
+        throwError "Do not yet support custom std_exact?/std_apply? tactics."
+    let allowFailure := fun g => do
+      let g ← g.withContext (instantiateMVars (.mvar g))
+      return required.all fun e => e.occurs g
+    match ← librarySearch goal tactic allowFailure with
+    -- Found goal that closed problem
+    | none =>
+      addExactSuggestion tk (← instantiateMVars (mkMVar mvar)).headBeta
+    -- Found suggestions
+    | some suggestions =>
+      if requireClose then throwError
+        "`std_exact?` could not close the goal. Try `std_apply?` to see partial suggestions."
+      reportOutOfHeartbeats `library_search tk
+      for (_, suggestionMCtx) in suggestions do
+        withMCtx suggestionMCtx do
+          addExactSuggestion tk (← instantiateMVars (mkMVar mvar)).headBeta (addSubgoalsMsg := true)
+      if suggestions.isEmpty then logError "std_apply? didn't find any relevant lemmas"
```

Comment:
```suggestion
      if suggestions.isEmpty then logError "apply? didn't find any relevant lemmas"
```

Reply:
Or even
```suggestion
      if suggestions.isEmpty then logError "exact? didn't find any relevant lemmas"
```

Reply:
Does this code path know which tactic we're in?

Reply:
It happens to be `apply?` due to the `requresClose` check above.  I documented and clarified.